### DISCUSSION TITLE
2.9 Plans and Proposals

### DISCUSSION THREAD

**Discussion by sovdeeth** [SkriptLang/Skript#6521] [General]
Hello all! I have been known to forget the details on exactly what us over on the SkriptLang team have discussed or planned for future versions, so I threw out the idea of creating a Github Discussion details our major plans for the next version. With any luck, this should serve as a reference piece for the team and any contributors about where work should be directed, and as a way for the community to be aware of the direction of the project and provide feedback and suggestions. If this post proves useful, I plan on continuing it for future versions.

With that said, here's a rundown on the major pushes for 2.9. Please remember all of this is subject to change, these are just goals, not hard deadlines.
### New Registration API:
@APickledWalrus and @kiip1 have been hard at work on a major rework to Skript's core interface, decoupling it further from Bukkit and providing new and improved methods to register syntax. This is a major step towards our end goal of becoming less tied to the Bukkit API while also providing a lot more control over possible syntax conflicts and priorities. If you're an addon developer, I highly suggest you check it out at <https://github.com/SkriptLang/Skript/pull/6246> and leave some feedback.

### Deprecation of Aliases
We're planning on moving toward removing or replacing the aliases system in 2.10, in some fashion. Aliases are one of the, in my opinion, least intuitive and most annoying aspects of Skript, given that they balloon startup times immensely for what seems to be little benefit. 

For 2.9, we don't intend to do anything drastic. Our current plans center around exposing the built-in aliases so that users can pick and choose which ones they would like to enable. Notably, as pointed out in <https://github.com/SkriptLang/skript-aliases/issues/111>, armor trims make up 15% of the aliases and are hardly ever used. Being able to disable those aliases from being generated would be a good first step.

Moving forward, our tentative plans are focusing on switching off the current aliases by default, leaving auto-generated aliases ("diamond sword" from "minecraft:diamond_sword") which will be a major change but one that should improve the intuitive-ness of the language. Blockdata will also fill the gaps where some of the more onerous aliases, like those for stairs, used to be. 

We'd love to hear more feedback about these plans, or other suggestions the community may have, in the comments of this issue.

### Better Variable Storage
@TheLimeGlass has been hard at work on getting SQL and H2 working as databases for variables. This has been a long time coming and should prove to be a major improvement to how Skript stores variables. If you're interested in testing this change out, there's a running test of it available on [our testing discord](https://discord.gg/ZPsZAg6ygu)! As we move to the final review stages of this addition, feedback and testing to make sure it's working as intended are very valuable, so please take some time to test it out if you're interested.

### Java 8 Support
Given that the vast majority of our userbase is on Java version 17 or above, and that modern versions of Minecraft and Paper no longer support Java 8, we're looking into dropping support for Java 8 and moving to a more modern standard. This will likely be Java 11 for 2.9, as Spigot 1.13 supports Java 11, and then Java 17 for 2.10. This would mean 2.10 would be dropping support for 1.15 and below. However, it would allow us to use a lot of very useful Java features we currently cannot take advantage of. Let us know your thoughts.

### Docs Tutorials
We've had the tutorials page on the docs website for years now, with only the text tutorial available. We're working on upgrading the website to be much more friendly to contributors who want to submit tutorials, and hopefully this post will remind us to keep working on the site to get it ready for release. 

If you're interested in contributing tutorials of your own, please check out <https://github.com/SkriptLang/skript-docs/pulls>. You can see my Command tutorial PR there as an example.

**Comment by EquipableMC** [comment]
I have a few more mentions that would really benefit us.

### Priority
With worldguard support for skript, it would make sense for Priority to actually work. I know there is code checking if the event is cancelled, but that doesn't work, which leads us to having to do an annoying workaround. It would be nice if this was fixed by 2.9.

### Text Displays
Text Displays would be beneficial to add into skript to make creating them easier since you can   Skript has support for armor stands.

These are just some of the proposed ideas that are very requested

**Reply by sovdeeth** [reply]
Luckily for you, event priorities already work! If you're talking about listening for cancelled events, that's already ready and will be out in 2.9.

Text displays are one thing I forgot to mention, but Lime's PR for them is nearly complete, and should be ready for 2.9 as well.

**Reply by EquipableMC** [reply]
I saw comments from you all saying how event priorities don't work because (as of right now) they are broken, unless I maybe misunderstood it?

**Reply by sovdeeth** [reply]
You likely misunderstood it, yes. Priorities work, but Skript currently can't listen to cancelled events, meaning they're not as useful as they could be.

**Comment by TheLimeGlass** [comment]
Just go to 3.0 not 2.10. Also sealed classes from Java 15 and SpigotLibraryLoader for Java 17 would be nice to utilize. Not Java 11.

**Reply by kiip1** [reply]
Version 3.0 should be reserved for a major Skript rewrite. I do agree with skipping Java 11 though, first and foremost due to the new features we would have access to, but also because servers stuck on older versions will probably not have the capability to move to Skript 2.9 anyways, this small percentage of servers should not hold us back from using a newer version.

**Reply by sovdeeth** [reply]
Good points, we'll definitely consider jumping straight to 17. It would be nice to provide warning a version in advance, hence the 11 then 17 plan, but we'll evaluate if that would even be worthwhile.
Definitely not calling it Skript 3, though.

**Comment by cheeezburga** [comment]
I think some of the most tedious things that users encounter are item comparison related. I know there were plans to rework this, but I haven't looked into it much since the conversation in #skriptlang-development. Would this be something for 2.9, or at a later date so users don't feel out of the loop and stuff?

**Reply by sovdeeth** [reply]
We're mainly waiting to see what the changes in 1.20.5 do to the item landscape in Bukkit/Paper. It could stay the same, or it could open up new opportunities for us. Not sure if it'll be in 2.9 or not.

**Comment by APickledWalrus** [comment]
2.9 is here: https://github.com/SkriptLang/Skript/releases/tag/2.9.0
